File: front-end/src/Pages/map-generator/procedural.js
const universe_variables = {
    min_tiles_from_edge: 3,
    map_width: 200,
    map_height: 200,
    chance_of_island_expansion: 55,
    map_coverage: 20,
    max_island_width: 500,
    max_island_height: 500,
    min_island_width: 20,
    min_island_height: 20,
}

const add_isle_to_map = (map, island) => {
    const island_height = island.length
    const island_width = island[0].length
    const map_height = map.length
    const map_width = map[0].length
    let start_x = Math.floor(Math.random()*(map_width-island_width))
    let start_y = Math.floor(Math.random()*(map_height-island_height))
    const max_tries = 1000
    let can_add = true
    let island_size = 0
    
    for(let tries=0; tries<max_tries; tries++) {
        for(let row=start_y; row<start_y+island_height; row++) {
            for(let col=start_x; col<start_x+island_width; col++) {
                if(island[row-start_y][col-start_x]) island_size++
                if(map[row][col] && island[row-start_y][col-start_x]) {
                    can_add = false
                    break
                }
            }
            if(!can_add) break
        }
        if(can_add) break
        island_size = 0
    }

    if(can_add) {
        for(let row=start_y; row<start_y+island_height; row++)
            for(let col=start_x; col<start_x+island_width; col++)
                if(island[row-start_y][col-start_x])
                    map[row][col] = island[row-start_y][col-start_x]
        return island_size
    } else return 0
}

const add_to_tileset = (tileset, new_tile) => {
    let index = null
    for(tileset_idx=0; tileset_idx<tileset.length; tileset_idx++) {
        if(JSON.stringify(tileset[tileset_idx]) === JSON.stringify(new_tile)) {
            index = tileset_idx; break
        }
    }
    if(index === null || tileset.length === 0) {
        index = tileset.length
        tileset.push(new_tile)
    }
    return index
}

const generate_map = (tileset, config) => {
    // const num_of_islands = Math.floor(config.map_width * config.map_height * config.island_tile_ratio)
    let island_tileset = []
    let map = []
    let map_coverage = 0
    for(let row=0;row<config.map_height;row++) map[row] = new Array(config.map_width).fill(null)

    console.log('...generating islands')
    while(map_coverage<config.map_height*config.map_width*(config.map_coverage/100)) {
        const island = generate_island(config, tileset, island_tileset)
        // console.log(island)
        //add island_tileset to map_tileset

        //add island to map
        map_coverage += add_isle_to_map(map, island)
    }
    
    //to optimiize this move it to island tiles
    console.log('...creating tileset')
    for(let row=0; row<map.length; row++) {
        for(let col=0; col<map[0].length; col++) {
            map[row][col] = add_to_tileset(island_tileset, map[row][col])
        }
    }

    console.log('...drawing map')


    return {tileset: island_tileset, map: map}
    // console.log(map)
}

File: temp--map_stuff/python/procedural.py
from random import randrange as rand
from tileset import tileset
from config import config
from generate_island import generate_island

def add_to_tileset(tileset, tile):
    for tile_index in range(len(tileset)):
        if tile == tileset[tile_index]: return tile_index
    tileset.append(tile)
    return len(tileset)

def add_to_map(map, island, island_tileset):
    map_height =  len(map)
    map_width = len(map[0])
    island_height = len(island)
    island_width = len(island[0])
    max_tries = 1
    island_size = 0

    for _ in range(max_tries):
        start_x = rand(map_width-island_width)
        start_y = rand(map_height-island_height)
        can_add = True
        for row in range(start_y, start_y+island_height, 1):
            for col in range(start_x, start_x+island_width, 1):
                if island[row-start_y][col-start_x] is not 0: island_size += 1
                if map[row][col] is not 0 and island[row-start_y][col-start_x] is not 0:
                    can_add = False
                    break
            if not can_add: break
        if can_add: break
    
    if can_add:
        for row in range(start_y, start_y+island_height, 1):
            for col in range(start_x, start_x+island_width, 1):
                if island[row-start_y][col-start_x] is not 0:
                    map[row][col] = add_to_tileset(island_tileset, island[row-start_y][col-start_x])
    
    return island_size


def generate_map(tileset, config):
    island_tileset = []
    map = [[0 for x in range(config['map']['width'])] for y in range(config['map']['width'])]
    map_coverage = 0
    while map_coverage < config['map']['width']*config['map']['height']*(config['map']['coverage']/100):
        island = generate_island(tileset['island'], config['island'])
        map_coverage += add_to_map(map, island, island_tileset)
    return {'tileset': island_tileset, 'map': map, 'colors': tileset['island']['colors']}


print(generate_map(tileset, config))